### ISSUE TITLE
A possible bug in full_split ?

### ISSUE THREAD

**Issue by chetmurthy** [mmottl/pcre-ocaml#29]
Over at the OCaml discussion forum, someone found what looks like a bug with `full_split`: https://discuss.ocaml.org/t/full-split-in-pcre/12948/1

TL;DR it looks like full_split isn't getting unmatched groups right for delimiters after the first delimiter.  I dug into the code, and it seems like the problem might be with the way that full_split invokes `unsafe_pcre_exec`.  To wit, everywhere else except in the `split` functions (full_split, internal_psplit) the value of `subj_start` is always `0`.  But in these two functions, it is the same as `pos`.

When I changed this to `0`, the bug evinced in the post above disappeared.

I'm going to dig further, but I thought I'd report this, and ask for your feedback.  If you're still actively maintaining this code, I'll send a PR; otherwise, I'll just fix it in the `pcre2` codebase.

**Comment by mmottl** [comment]
Thanks a lot for the bug report and sorry for the delay. This issue revealed superfluous code (almost certainly - it's ancient!) that also caused a subtle bug due to the way PCRE reports non-capturing groups. Please feel free to test the latest development version, it should handle such groups in full splits correctly now.

**Comment by chetmurthy** [comment]
I'll give it a try and report back if I find any problems.  Do you have any idea when you might release ?  Just so I can know when I should be thinking of updating my code that currently works-around this?

**Comment by chetmurthy** [comment]
Markus, I tried your patched tree on my existing code that uses pcre, and got a core-dump.  I haven't started debugging the source of the crash, but I figured I should let you know (you might have more knowledge at-hand and be able to diagnose more quickly).  Unless I hear back from you, I'll proceed with debugging pcre-ocaml, and maybe I'll be able to report what went wrong.

(1) I tested on ocaml versions 4.14.0, 5.0.0, 5.1.0, 5.2.0

(2) I installed with "opam install ." right from your source-tree

(3) and here's the small testcase:

Two files: buzz.ml
```
open Pcre ;;
let re = Pcre.regexp ~flags:[] "^#!(\\S+)" ;;
let f subj = Pcre.exec ~rex:re subj ;;

f "#!/foo/bar/buzz" ;;
```
and: DOIT
```
#!/bin/bash -x

rm -f buzz.cm* buzz.exe
ocamlfind ocamlc -linkall -linkpkg  -package  pcre -linkpkg -linkall -o buzz.exe buzz.ml
./buzz.exe
```

Running `DOIT` produces
```
+ rm -f buzz.cmi buzz.cmo buzz.exe
+ ocamlfind ocamlc -linkall -linkpkg -package pcre -linkpkg -linkall -o buzz.exe buzz.ml
+ ./buzz.exe
./DOIT: line 5: 84127 Segmentation fault      (core dumped) ./buzz.exe

**Comment by chetmurthy** [comment]
A little more: here are two commits (from pcre-ocaml):
```
efc1044 * Removed superfluous and buggy subj_start argument
795239b * Updated GitHub workflow
```

The later one elicits the bug; the earlier one does not.

**Comment by chetmurthy** [comment]
Also: I'm not going to pretend that I can be efficient at debugging this problem (it's been a few years since the last time I debugged in OCaml<->C code), so I'll just report what I find and hope it speeds up your debugging.  In that spirit:

I found that switching to the opt compiler made the above testcase work without problem.  I haven't gone back thru all the other examples (from which I narrowed-down this testcase) to see if they also worked with the opt compiler.

**Comment by mmottl** [comment]
Thanks for testing! Though I found a tangential bug, the problem is actually that the byte code dynamically links with the PCRE C-stubs installed in OPAM, not the one in the tree. Since the C-stubs have changed, we get a segfault.

The solution is to pin the package to the development directory using `opam pin .` before testing. Then both native code and byte code should work (certainly with the latest fix for the tangential bug).

**Comment by chetmurthy** [comment]
Um, I don't think that's the problem.  I was pretty careful to verify that I only had one version of pcre installed, using this script to reinstall from your source-tree:
```
#!/bin/bash -x

opam pin remove pcre
opam remove pcre
opam install --working-dir .
```
The install pins pcre to the source-dir, viz.
```
$ opam pin list
pcre.7.5.0    git  git+file:///home/chet/Hack/Camlp5/src/ALL/pcre-ocaml#master
```

**Comment by mmottl** [comment]
The latest [fix](https://github.com/mmottl/pcre-ocaml/commit/c5b8afb76bf13f9b759d32954a9b7e36b8339d7c)  should have addressed your problem. Your example works for me for both native code and byte code now. Is it still failing? When testing with byte code, it's still important to make sure any older `pcre-ocaml` cannot interfere.

**Comment by chetmurthy** [comment]
aha, yes, the latest fix makes the testcase work.  I will now test with the original example that caused the problem and report back.

**Comment by chetmurthy** [comment]
Excellent!  That worked!  I'll now move on to testing your full_split fixes!  Thank you!

**Comment by chetmurthy** [comment]
Right, that seems to have done the trick: I get comprehensible results from full_split.  I'll get busy integrating support for this into `pa_ppx_regexp`.  Thank you for this!

I confess that though I'm ostensibly the maintainer of pcre2-ocaml, I haven't had a look down into the guts to see how it works.  So if you do end up (as you mentioned) looking into fixing it there too, I'd appreciate it.  But otherwise, well, eventually I'll have a look at your changes and see if I can figure out how to do similar things to pcre2-ocaml.

**Comment by mmottl** [comment]
It shouldn't be hard to merge my latest changes. I'll send you a PR once I'm done.

**Comment by mmottl** [comment]
It seems nothing needs to be done for the newer PCRE2-OCaml since it is a more or less complete rewrite. I'll make a (last?) release of PCRE-OCaml with the fixed `full_split` function.
